File: languages/php/tests/LogCaptureTest.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Tests;

use PHPUnit\Framework\TestCase;
use Revaly\Sdk\Core\Model\CreditCard;
use Revaly\Sdk\Core\Model\PaymentMethod;
use Revaly\Sdk\Core\Model\PaymentRequest;
use Revaly\Sdk\Errors\OutcomeUnknownException;
use Revaly\Sdk\Errors\PermanentRejectionException;
use Revaly\Sdk\Logging\RapWireTrace;
use Revaly\Sdk\Testing\RapMockTransport;
use Revaly\Sdk\Testing\SyntheticData;
use Revaly\Sdk\Tests\Support\CollectingLogger;
use Revaly\Sdk\Tests\Support\TestClients;

/**
 * The ADR-SDK-020 log-capture obligations (DX contract §c): default output is
 * values-free; debug level carries allowlist-scrubbed payloads only; the correlation id
 * is on every error path; the wire-trace hook never sees raw material.
 */
final class LogCaptureTest extends TestCase
{
    /** A charge request carrying synthetic card material that must never be logged. */
    private static function cardChargeRequest(): PaymentRequest
    {
        $card = new CreditCard();
        $card->setNumber(TestClients::SYNTHETIC_PAN);
        $card->setCardVerificationCode(TestClients::SYNTHETIC_CVV);

        $method = new PaymentMethod();
        $method->setCreditCard($card);

        $request = TestClients::chargeRequest();
        $request->setPaymentMethodType('creditCard');
        $request->setPaymentMethod($method);

        return $request;
    }

    public function testDefaultOutputIsValuesFreeOnSuccess(): void
    {
        $logger = new CollectingLogger();
        $mock = new RapMockTransport();
        $mock->charge()->returnsApproved();

        TestClients::withMock($mock, logger: $logger)->charge(self::cardChargeRequest());

        $haystack = $logger->flattened();
        self::assertNotSame('', $haystack);
        self::assertStringNotContainsString(TestClients::SYNTHETIC_PAN, $haystack);
        self::assertStringNotContainsString(TestClients::API_KEY, $haystack);

        $info = $logger->atLevel('info');
        self::assertNotEmpty($info);
        self::assertSame('charge', $info[0]['context']['operation']);
        self::assertSame(200, $info[0]['context']['status']);
        self::assertSame(SyntheticData::DEFAULT_CORRELATION_ID, $info[0]['context']['correlation']);
    }

    public function testFailurePathLogsClassAndCorrelationOnly(): void
    {
        $logger = new CollectingLogger();
        $mock = new RapMockTransport();
        $mock->charge()->returnsPermanentRejection(422);

        try {
            TestClients::withMock($mock, logger: $logger)->charge(self::cardChargeRequest());
            self::fail('expected PermanentRejectionException');
        } catch (PermanentRejectionException $expected) {
        }

        $haystack = $logger->flattened();
        self::assertStringNotContainsString(TestClients::SYNTHETIC_PAN, $haystack);
        self::assertStringNotContainsString(TestClients::API_KEY, $haystack);

        $warnings = $logger->atLevel('warning');
        self::assertNotEmpty($warnings);
        self::assertSame('PermanentRejection', $warnings[0]['context']['class']);
        // Correlation id on every error path (DX contract §c).
        self::assertSame(SyntheticData::DEFAULT_CORRELATION_ID, $warnings[0]['context']['correlation']);
    }

    public function testDebugPayloadsAreScrubbed(): void
    {
        $logger = new CollectingLogger();
        $mock = new RapMockTransport();
        $mock->charge()->returnsApproved();

        TestClients::withMock($mock, logger: $logger)->charge(self::cardChargeRequest());

        $debug = $logger->atLevel('debug');
        self::assertNotEmpty($debug);
        $request = (string) $debug[0]['context']['request'];
        self::assertStringContainsString('[scrubbed]', $request);
        self::assertStringContainsString('mtx-synthetic-1', $request);
        self::assertStringNotContainsString(TestClients::SYNTHETIC_PAN, $request);
    }

    public function testWireTraceHookReceivesOnlyScrubbedPayloads(): void
    {
        /** @var list<RapWireTrace> $traces */
        $traces = [];
        $mock = new RapMockTransport();
        $mock->charge()->returnsApproved();

        TestClients::withMock(
            $mock,
            wireTraceHook: function (RapWireTrace $trace) use (&$traces): void {
                $traces[] = $trace;
            },
        )->charge(self::cardChargeRequest());

        self::assertCount(1, $traces);
        $trace = $traces[0];
        self::assertSame('charge', $trace->operation);
        self::assertSame('POST', $trace->method);
        self::assertSame(200, $trace->status);
        self::assertSame(SyntheticData::DEFAULT_CORRELATION_ID, $trace->correlationId);
        self::assertStringNotContainsString(TestClients::SYNTHETIC_PAN, (string) $trace->scrubbedRequestBody);
        self::assertStringContainsString('[scrubbed]', (string) $trace->scrubbedRequestBody);
        self::assertStringContainsString('mtx-synthetic-1', (string) $trace->scrubbedRequestBody);
    }

    public function testWireTraceHookExceptionsAreSwallowed(): void
    {
        $mock = new RapMockTransport();
        $mock->charge()->returnsApproved();

        $transaction = TestClients::withMock(
            $mock,
            wireTraceHook: static function (): void {
                throw new \RuntimeException('observer bug');
            },
        )->charge(self::cardChargeRequest());

        // Tracing must never change payment control flow (runtime-tdd §6).
        self::assertSame(1, $transaction->getTransactionStatus());
    }

    public function testReconcileTraceCarriesScrubbedResponseOnly(): void
    {
        /** @var list<RapWireTrace> $traces */
        $traces = [];
        $mock = new RapMockTransport();
        $mock->reconcile(SyntheticData::DEFAULT_MERCHANT_TRANSACTION_ID)->returnsApproved();

        TestClients::withMock(
            $mock,
            wireTraceHook: function (RapWireTrace $trace) use (&$traces): void {
                $traces[] = $trace;
            },
        )->reconcile(
            SyntheticData::DEFAULT_MERCHANT_TRANSACTION_ID,
            new \Revaly\Sdk\Reconcile\ReconcilePolicy(1, 5.0, 0.0),
        );

        self::assertCount(1, $traces);
        self::assertSame('reconcile', $traces[0]->operation);
        self::assertStringContainsString('[scrubbed]', (string) $traces[0]->scrubbedResponseBody);
        self::assertStringContainsString('mtx-synthetic-1', (string) $traces[0]->scrubbedResponseBody);
    }

    public function testTransportFailureLogsNeverContainTheApiKey(): void
    {
        $logger = new CollectingLogger();
        $mock = new RapMockTransport();
        $mock->charge()->throwsTimeoutAfterSend();

        try {
            TestClients::withMock($mock, logger: $logger)->charge(self::cardChargeRequest());
            self::fail('expected OutcomeUnknownException');
        } catch (OutcomeUnknownException $expected) {
        }

        self::assertStringNotContainsString(TestClients::API_KEY, $logger->flattened());
    }
}

File: languages/go/tests/logcapture_test.go
// ADR-SDK-020 log-capture suite (DX contract §c acceptance): run a charge, a
// failure, and a reconcile against the mock transport with logging at debug
// level and a wire-trace hook attached, then assert no PAN/CVV/key material
// anywhere — logs, trace events, or error messages — and correlation ids on
// every error path. Plus direct units for the central scrubber.
package tests

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"strings"
	"sync"
	"testing"

	revaly "github.com/revaly-co/rap-sdk/languages/go"
	core "github.com/revaly-co/rap-sdk/languages/go/core"
	"github.com/revaly-co/rap-sdk/languages/go/internal/runtime"
)

const (
	syntheticPAN = "4111111111111111"
	syntheticCVV = "9876"
)

// capturingHandler renders every record (message + attrs) into a string list.
type capturingHandler struct {
	mu    sync.Mutex
	lines *[]string
	attrs []slog.Attr
}

func (h *capturingHandler) Enabled(context.Context, slog.Level) bool { return true }

func (h *capturingHandler) Handle(_ context.Context, record slog.Record) error {
	var b strings.Builder
	b.WriteString(record.Message)
	for _, attr := range h.attrs {
		fmt.Fprintf(&b, " %s=%v", attr.Key, attr.Value)
	}
	record.Attrs(func(attr slog.Attr) bool {
		fmt.Fprintf(&b, " %s=%v", attr.Key, attr.Value)
		return true
	})
	h.mu.Lock()
	*h.lines = append(*h.lines, b.String())
	h.mu.Unlock()
	return nil
}

func (h *capturingHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	return &capturingHandler{lines: h.lines, attrs: append(append([]slog.Attr{}, h.attrs...), attrs...)}
}

func (h *capturingHandler) WithGroup(string) slog.Handler { return h }

func cardChargeRequest() *core.PaymentRequest {
	card := core.NewCreditCard(syntheticPAN, "12", "2030")
	card.SetCardVerificationCode(syntheticCVV)
	method := core.NewPaymentMethod()
	method.SetCreditCard(*card)
	request := core.NewPaymentRequest(1099, "mock-mtx-log-1")
	request.SetPaymentMethod(*method)
	return request
}

func TestValuesFreeLoggingAndScrubbedTraces(t *testing.T) {
	var lines []string
	var events []revaly.WireTraceEvent
	var eventsMu sync.Mutex

	client, mock := newMockClient(t, func(cfg *revaly.Config) {
		cfg.Logger = slog.New(&capturingHandler{lines: &lines})
		cfg.WireTrace = func(event revaly.WireTraceEvent) {
			eventsMu.Lock()
			events = append(events, event)
			eventsMu.Unlock()
		}
	})

	// charge (success) + charge (failure) + reconcile — the §c acceptance flow.
	mock.Charge().ReturnsApproved().ReturnsNotProcessed503()
	mock.Reconcile("mock-mtx-log-1").ReturnsNotFound().ReturnsFoundApproved()

	if _, err := client.Charge(context.Background(), cardChargeRequest()); err != nil {
		t.Fatalf("charge 1: %v", err)
	}
	_, chargeErr := client.Charge(context.Background(), cardChargeRequest())
	var transient *revaly.TransientFailure
	if !errors.As(chargeErr, &transient) {
		t.Fatalf("charge 2: got %v, want *TransientFailure", chargeErr)
	}
	if _, err := client.Reconcile(context.Background(), "mock-mtx-log-1", policy(3)); err != nil {
		t.Fatalf("reconcile: %v", err)
	}

	joinedLogs := strings.Join(lines, "\n")
	var traceBodies strings.Builder
	eventsMu.Lock()
	for _, event := range events {
		traceBodies.WriteString(event.ScrubbedRequestBody)
		traceBodies.WriteString(event.ScrubbedResponseBody)
	}
	eventsMu.Unlock()

	for name, corpus := range map[string]string{
		"logs":          joinedLogs,
		"trace bodies":  traceBodies.String(),
		"error message": chargeErr.Error(),
	} {
		for _, secret := range []string{syntheticPAN, syntheticCVV, syntheticKey} {
			if strings.Contains(corpus, secret) {
				t.Fatalf("%s contain sensitive material %q:\n%s", name, secret, corpus)
			}
		}
	}

	if len(lines) == 0 {
		t.Fatal("expected values-free log lines")
	}
	if !strings.Contains(joinedLogs, "correlation=mock-corr-") {
		t.Fatalf("error-path logs missing the correlation id:\n%s", joinedLogs)
	}
	if transient.CorrelationID == "" {
		t.Fatal("typed error missing the correlation id")
	}

	// The trace hook saw scrubbed material: allowlisted identifiers pass,
	// payload values do not.
	eventsMu.Lock()
	defer eventsMu.Unlock()
	if len(events) == 0 {
		t.Fatal("expected wire-trace events")
	}
	sawChargeRequest := false
	for _, event := range events {
		if event.Operation == "charge" && strings.Contains(event.ScrubbedRequestBody, "mock-mtx-log-1") {
			sawChargeRequest = true
		}
	}
	if !sawChargeRequest {
		t.Fatal("trace events missing the allowlisted merchantTransactionId")
	}
}

// A panicking trace hook must never affect payment control flow.
func TestWireTraceHookPanicsAreSwallowed(t *testing.T) {
	client, mock := newMockClient(t, func(cfg *revaly.Config) {
		cfg.WireTrace = func(revaly.WireTraceEvent) { panic("observer bug") }
	})
	mock.Charge().ReturnsApproved()
	if _, err := client.Charge(context.Background(), chargeRequest("mock-mtx-0001")); err != nil {
		t.Fatalf("charge with a panicking hook: %v", err)
	}
}

func TestScrubberAllowlist(t *testing.T) {
	scrubbed := runtime.ScrubJSON(`{"merchantTransactionId":"mtx-1","number":"` + syntheticPAN + `",` +
		`"cardVerificationCode":"` + syntheticCVV + `","transactionStatus":1,` +
		`"nested":{"code":"not_processed","customerName":"Jane Cardholder"}}`)
	for _, keep := range []string{`"merchantTransactionId":"mtx-1"`, `"transactionStatus":1`, `"code":"not_processed"`} {
		if !strings.Contains(scrubbed, keep) {
			t.Fatalf("allowlisted field lost: %s missing from %s", keep, scrubbed)
		}
	}
	for _, gone := range []string{syntheticPAN, syntheticCVV, "Jane Cardholder"} {
		if strings.Contains(scrubbed, gone) {
			t.Fatalf("sensitive value survived the scrub: %s in %s", gone, scrubbed)
		}
	}
	if !strings.Contains(scrubbed, runtime.Scrubbed) {
		t.Fatalf("expected scrub tokens in %s", scrubbed)
	}
}

func TestScrubberNonJSONNeverEchoes(t *testing.T) {
	if got := runtime.ScrubJSON("PAN " + syntheticPAN + " raw text"); strings.Contains(got, syntheticPAN) {
		t.Fatalf("raw text echoed: %s", got)
	}
}

func TestScrubberHeadersRedactAuthorization(t *testing.T) {
	headers := http.Header{}
	headers.Set("Authorization", "ApiKey "+syntheticKey)
	headers.Set("Content-Type", "application/json")
	headers.Set("X-Correlation-ID", "corr-1")
	headers.Set("X-Custom", "value")
	scrubbed := runtime.ScrubHeaders(headers)
	if scrubbed["Authorization"] != runtime.Redacted {
		t.Fatalf("Authorization = %q, want redacted", scrubbed["Authorization"])
	}
	if strings.Contains(fmt.Sprint(scrubbed), syntheticKey) {
		t.Fatal("API key survived header scrubbing")
	}
	// http.Header canonicalizes stored keys (X-Correlation-ID → X-Correlation-Id).
	if scrubbed["Content-Type"] != "application/json" || scrubbed[http.CanonicalHeaderKey("X-Correlation-ID")] != "corr-1" {
		t.Fatalf("allowlisted headers mangled: %v", scrubbed)
	}
	if scrubbed["X-Custom"] != runtime.Redacted {
		t.Fatalf("unknown header not redacted: %v", scrubbed)
	}
}
